src/rez/tests/test_test.py

```diff
@@ -0,0 +1,323 @@
+                packages_path=[self._install_path],
+                variable="REZ_CURRENT_TEST_NAME",
+            )
+
+    def test_conflicting_test_requires(self):
+        """Fail to resolve because the test requires conflicts with package requires."""
+        with self.assertRaises(rez_exceptions.PackageTestError):
+            _test_shell(
+                "test package_with_tests --interactive invalid_test",
+                packages_path=[self._install_path],
+                variable="REZ_RESOLVED_WITH_TESTS",
+            )
+
+    def test_missing_argument(self):
+        """Fail to run if ``--interactive`` is given without a test name."""
+        with self.assertRaises(SystemExit) as found:
+            _test_shell(
+                "test package_with_tests --interactive",
+                packages_path=[self._install_path],
+                variable="REZ_RESOLVED_WITH_TESTS",
+            )
+
+        self.assertEqual(2, found.exception.code)
+
+    def test_not_found_test_name(self):
+        """Fail if the given test name isn't defined in the package."""
+        with self.assertRaises(rez_exceptions.PackageTestError):
+            _test_shell(
+                "test package_with_tests --interactive does_not_exist",
+                packages_path=[self._install_path],
+                variable="REZ_RESOLVED_WITH_TESTS",
+            )
+
+
+@contextlib.contextmanager
+def _override_execute_shell(variable="REZ_RESOLVE"):
+    """Change Rez's ResolvedContext.execute_shell to print and return.
+
+    Args:
+        variable (str, optional):
+            An environment variable to print and return as stdout.
+
+    Yields:
+        callable:
+            The overwritten
+            :meth:`rez.resolved_context.ResolvedContext.execute_shell` method.
+
+    """
+    _, path = tempfile.mkstemp(suffix="_override_execute_shell.txt")
+    atexit.register(functools.partial(os.remove, path))
+
+    def _simulate_non_interactive(kwargs):
+        system = platform.system()
+        output = kwargs.copy()
+
+        if system == "Windows":
+            output["command"] = "echo %{variable}% >> {path}".format(variable=variable, path=path)
+        else:
+            output["command"] = "echo ${variable} >> {path}".format(variable=variable, path=path)
+
+        output["block"] = True
+
+        return output
+
+    def _wrap(function):
+        @functools.wraps(function)
+        def wrapper(*args, **kwargs):
+            """Append a command to print the resolved packages, then run ``function``."""
+            kwargs = _simulate_non_interactive(kwargs)
+
+            return function(*args, **kwargs)
+
+        return wrapper
+
+    original = resolved_context.ResolvedContext.execute_shell
+
+    try:
+        resolved_context.ResolvedContext.execute_shell = _wrap(
+            resolved_context.ResolvedContext.execute_shell
+        )
+
+        yield path
+    finally:
+        resolved_context.ResolvedContext.execute_shell = original
```

Comment:
This code is fairly hacky but is necessary to prevent interactive from **actually** rez-enving, because we don't actually want that to happen in the unittest. If we had `mock` as a vendor, I would've `mock.patch`ed here. But wasn't sure if there that's okay to add. Just wanted to raise here that `mock` would've been helpful here, for what it's worth.

Reply:
If it makes our life easier and the code much easier to read and understand, we should really consider adding `mock` as a runtime dependency. I don't think we need to vendor it...

Reply:
I thought Rez had to copy all external dependencies into the vendor folder - Something Allan mentioned on the forums about Rez needing to be self enclosed. Is that no longer the case?

Reply:
To me test dependencies should not br vendored.

We could easily have an extra in the setup.py that installs pytest and mock.

Reply:
Oh okay cool, I didn't know that was an option. I'm happy to do that if others agree as well!